src/chart/context.rs

```diff
@@ -549,81 +516,70 @@ impl<'a, DB: DrawingBackend, X: Ranged, Y: Ranged> ChartContext<'a, DB, RangedCo
-                .unwrap_or((0, 0));
-
-            let (cx, cy) = if tick_size >= 0 {
+            let (cx, cy, h_pos, v_pos) = if tick_size >= 0 {
                 match orientation {
                     // Right
                     (dx, dy) if dx > 0 && dy == 0 => {
-                        (right_alignment + label_dist - w as i32, *p - y0)
+                        (label_dist, *p - y0, HPos::Left, VPos::Center)
                     }
                     // Left
-                    (dx, dy) if dx < 0 && dy == 0 => (tw as i32 - label_dist - w as i32, *p - y0),
+                    (dx, dy) if dx < 0 && dy == 0 => {
+                        (tw as i32 - label_dist, *p - y0, HPos::Right, VPos::Center)
+                    }
                     // Bottom
-                    (dx, dy) if dx == 0 && dy > 0 => (*p - x0, label_dist + h as i32),
+                    (dx, dy) if dx == 0 && dy > 0 => (*p - x0, label_dist, HPos::Center, VPos::Top),
                     // Top
-                    (dx, dy) if dx == 0 && dy < 0 => (*p - x0, th as i32 - label_dist - h as i32),
+                    (dx, dy) if dx == 0 && dy < 0 => {
+                        (*p - x0, th as i32 - label_dist, HPos::Center, VPos::Bottom)
+                    }
                     _ => panic!("Bug: Invalid orientation specification"),
                 }
             } else {
                 match orientation {
                     // Right
                     (dx, dy) if dx > 0 && dy == 0 => {
-                        (tw as i32 - right_alignment - label_dist, *p - y0)
+                        (tw as i32 - label_dist, *p - y0, HPos::Right, VPos::Center)
                     }
                     // Left
-                    (dx, dy) if dx < 0 && dy == 0 => (label_dist, *p - y0),
+                    (dx, dy) if dx < 0 && dy == 0 => {
+                        (label_dist, *p - y0, HPos::Left, VPos::Center)
+                    }
                     // Bottom
-                    (dx, dy) if dx == 0 && dy > 0 => (*p - x0, th as i32 - label_dist - h as i32),
+                    (dx, dy) if dx == 0 && dy > 0 => {
+                        (*p - x0, th as i32 - label_dist, HPos::Center, VPos::Bottom)
+                    }
                     // Top
-                    (dx, dy) if dx == 0 && dy < 0 => (*p - x0, label_dist + h as i32),
+                    (dx, dy) if dx == 0 && dy < 0 => (*p - x0, label_dist, HPos::Center, VPos::Top),
                     _ => panic!("Bug: Invalid orientation specification"),
                 }
             };
 
-            let should_draw = if orientation.0 == 0 {
-                cx >= 0 && cx + label_offset + w as i32 / 2 <= tw as i32
+            let (text_x, text_y) = if orientation.0 == 0 {
+                (cx + label_offset, cy)
             } else {
-                cy >= 0 && cy + label_offset + h as i32 / 2 <= th as i32
+                (cx, cy + label_offset)
             };
 
-            if should_draw {
```

Comment:
And one reason for this `should_draw`check is we need to avoid the tag is drawn out of it's own area, but if we only relies on the backend, a super long label may overlap with other drawing area on the right side (for example). And I think we should avoid this

Reply:
It could be confusing to users - if super long label is not drawn at all, they may think it's a bug in their code. I think the proper fix would be implement clipping in the drawing area, but I'm not sure how to properly do it with the current architecture (as drawing backend is not supposed to know about the drawing area). I guess it's possible to add a new API call for the backend like `set_clipping_region`, but it's rather high-level change that I won't risk to add without your explicit approval.

Reply:
Ok, I think your comment make sense. I think I am Ok with this . But I don't think we don't want the backends aware of any high-level details. Since I feel we should keep high-level and low-level API separate. `set_clipping_region` solution might be some solution, but  I think it needs some amount of work to do and I don't think it's worth us to dig into this.